rust/src/dumpfile.rs

```diff
@@ -0,0 +1,391 @@
+                    size,
+                    nlink,
+                    inline_content: content.map(unescape).transpose()?,
+                    fsverity_digest: fsverity_digest.map(ToOwned::to_owned),
+                },
+                libc::S_IFLNK => {
+                    let target =
+                        unescape_to_path(payload.ok_or_else(|| anyhow!("Missing payload"))?)?;
+                    Item::Symlink { nlink, target }
+                }
+                libc::S_IFIFO => Item::Fifo { nlink },
+                libc::S_IFCHR | libc::S_IFBLK => Item::Device { nlink, rdev },
+                libc::S_IFDIR => Item::Directory { size, nlink },
+                o => {
+                    anyhow::bail!("Unhandled mode {o:o}")
+                }
+            }
+        };
+        Ok(Entry {
+            path,
+            uid,
+            gid,
+            mode,
+            mtime,
+            item,
+            xattrs,
+        })
+    }
+
+    /// Remove internal entries
+    /// FIXME: This is arguably a composefs-info dump bug?
+    pub fn filter_special(mut self) -> Self {
+        self.xattrs.retain(|v| match (v.key.as_bytes(), &*v.value) {
+            (b"trusted.overlay.opaque" | b"user.overlay.opaque", b"x") => false,
+            _ => true,
+        });
+        self
+    }
+}
+
+impl<'p> Item<'p> {
+    pub(crate) fn size(&self) -> u64 {
+        match self {
+            Item::Regular { size, .. } | Item::Directory { size, .. } => *size,
+            _ => 0,
+        }
+    }
+
+    pub(crate) fn nlink(&self) -> u32 {
+        match self {
+            Item::Regular { nlink, .. } => *nlink,
+            Item::Device { nlink, .. } => *nlink,
+            Item::Symlink { nlink, .. } => *nlink,
+            Item::Directory { nlink, .. } => *nlink,
+            Item::Fifo { nlink, .. } => *nlink,
+            _ => 0,
+        }
+    }
+
+    pub(crate) fn rdev(&self) -> u32 {
+        match self {
+            Item::Device { rdev, .. } => *rdev,
+            _ => 0,
+        }
+    }
+
+    pub(crate) fn payload(&self) -> Option<&Path> {
+        match self {
+            Item::Symlink { target, .. } => Some(target),
+            Item::Hardlink { target } => Some(target),
+            _ => None,
+        }
+    }
+}
+
+impl Display for Mtime {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}.{}", self.sec, self.nsec)
+    }
+}
+
+impl<'p> Display for Entry<'p> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        escape(f, self.path.as_os_str().as_bytes())?;
+        write!(
+            f,
+            " {} {:o} {} {} {} {} {} ",
+            self.item.size(),
+            self.mode,
+            self.item.nlink(),
+            self.uid,
+            self.gid,
+            self.item.rdev(),
+            self.mtime,
+        )?;
+        if let Some(payload) = self.item.payload() {
+            write!(f, "{payload:?} ")?;
```

Comment:
Don't you need to escape here? Also, you need a custom escape handler for if payload is exactly "-" (to avoid it being parsed as None.